### ISSUE TITLE
PHP FPM jailed server mysql problem

### ISSUE THREAD

**Issue by SERBice** [virtualmin/virtualmin-gpl#570]
When moving php from FCGId mode to FPM on a jailed virtual server, PHP cannot connect to the mysql/mariadb socket.

I have tried removing the jail and putting it back, and it didn't work.

As far as I could observe it is because FPM is running inside the jail, and the mysql socket is not in the jail, so php has no access.

So, doing a bind of the socket inside the jail dir of the server solves it.
```
mount --bind /var/run/mysqld /home/chroot/21398729836/var/run/mysqld
```
(in my case I put it in fstab to keep it permanently)

In conclusion, with PHP FPM on a jailed server, the mysql socket is not accessible and therefore php cannot connect to the database (at least in my experience).

**Comment by jcameron** [comment]
Could you also potentially work around this by having PHP connect to MySQL via a TCP connection rather than a socket file?

**Comment by iliajie** [comment]
> In conclusion, with PHP FPM on a jailed server, the mysql socket is not accessible

What Virtualmin and Webmin version do you have installed on the server where you're seeing this issue?

**Comment by SERBice** [comment]
First of all, thank you for your answers.

I decided not to connect via tcp because I suspect a socket might be a better option (honestly I didn't even try to test tcp). I think if it was running in fcgid mode it should be able to switch to fpm without having to change other settings manually, that's why I'm reporting it here.

In a virtual server already created, with jail, I switched from fcgid (because it was consuming too much ram) to fpm, and that's when php could not connect to the mysql socket. I did a mount bind of the socket inside the jail and problem solved. Question, shouldn't this be done automatically by the virtualmin panel?


Ilia, I have the following versions (the server is a few days old, it is all new, it is not an old server that has been upgraded, it is a clean install):

Dist: Ubuntu Linux 22.04.2
Kernel: Linux 5.15.0-73-generic on x86_64
Webmin version 2.021
Usermin version 1.861
Virtualmin version 7.7
PHP Versions 8.1.18, 8.2.6
MySQL Ver 15.1 Distrib 10.6.12-MariaDB, for debian-linux-gnu (x86_64) using EditLine wrapper

**Comment by jcameron** [comment]
So I had another look into this, and I realized that Virtualmin already links the MariaDB socket file into the jail! 

If you check your MariaDB config file (typically `/etc/my.cnf` or files under `/etc/my.cnd.d`), is there a `socket=` line?

**Comment by SERBice** [comment]
Yes, the socket line is:
```
socket = /run/mysqld/mysqld.sock
```

And as I said before, I should bind /run/mysqld/ in the jail to access this socket, otherwise /home/chroot/21398729836/var/run/mysqld does not exist or was empty (I don't remember which one)

```
mount --bind /var/run/mysqld /home/chroot/21398729836/var/run/mysqld
```


In my case, the socket did not exist in the jail, I had to do it myself. 



--------------------
Additionally (and I think somewhat related to this jail issue), I encountered another problem, also related to the jail and the use of php in FPM mode, and that is that PHP cannot access the SMTP host and fails to make use of stream_socket_enable_crypto()

```
ErrorInfo: SMTP Error: Could not connect to SMTP host. Connection failed. stream_socket_enable_enable_crypto(): SSL operation failed with code 1. OpenSSL Error messages:
error:16000069:STORE routines::unregistered scheme
error:80000002:system library::No such file or directory
error:16000069:STORE routines::unregistered scheme
error:80000002:system library::No such file or directory
error:16000069:STORE routines::unregistered scheme
error:80000002:system library::No such file or directory
error:16000069:STORE routines::unregistered scheme
error:80000002:system library::No such file or directory
error:0A000086:SSL routines::certificate verify failedSMTP server error: QUIT command failed Detail: G�*�!�~�#~~{�
����gC����+���



(THE LOG CONTINUES, IT CONTAINS HOSTNAMES AND OTHER SENSITIVE INFORMATION, BUT THE IMPORTANT PART IS HERE)
```

This error only occurs with: a virtual server jailed and using php in FPM mode, by changing either of the two options the error is not present (FGCId and JAIL, or Non-Jail and FPM).

**Comment by jcameron** [comment]
A bind mount will work - but if you create a new test domain and enable chroot for it, is a hard link to that socket file created in /var/run/mysqld or /run/mysqld ?

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
